code/modules/security levels/keycard authentication.dm

```diff
@@ -1,145 +1,99 @@
-	var/mob/event_triggered_by
-	var/mob/event_confirmed_by
-	//1 = select event
-	//2 = authenticate
 	anchored = 1
 	use_power = 1
 	idle_power_usage = 2
 	active_power_usage = 6
 	power_channel = ENVIRON
+	req_access = list(access_keycard_auth)
+	var/datum/event/ev
+	var/event = ""
+	var/obj/machinery/keycard_auth/event_source
+	var/mob/triggerer = null
+	var/waiting = 0
 
-/obj/machinery/keycard_auth/attack_ai(mob/user)
-	if(IsAdminGhost(user))
-		return
-	user << "The station AI is not to interact with these devices."
-	return
+/obj/machinery/keycard_auth/New()
+	. = ..()
+	ev = keycard_events.addEvent("triggerEvent", src, "triggerEvent")
 
-/obj/machinery/keycard_auth/attack_paw(mob/user)
-	user << "<span class='warning'>You are too primitive to use this device!</span>"
-	return
+/obj/machinery/keycard_auth/Destroy()
+	keycard_events.clearEvent("triggerEvent", ev)
+	qdel(ev)
+	. = ..()
 
 /obj/machinery/keycard_auth/attackby(obj/item/weapon/W, mob/user, params)
-	if(stat & (NOPOWER|BROKEN))
-		user << "This device is not powered."
-		return
-	if(istype(W,/obj/item/weapon/card/id))
-		var/obj/item/weapon/card/id/ID = W
-		if(access_keycard_auth in ID.access)
-			if(active == 1)
-				//This is not the device that made the initial request. It is the device confirming the request.
-				if(event_source)
-					event_source.confirmed = 1
-					event_source.event_confirmed_by = usr
-			else if(screen == 2)
-				event_triggered_by = usr
-				broadcast_request() //This is the device making the initial event request. It needs to broadcast to other devices
-
-/obj/machinery/keycard_auth/power_change()
-	if(powered(ENVIRON))
-		stat &= ~NOPOWER
-		icon_state = "auth_off"
-	else
-		stat |= NOPOWER
-
-/obj/machinery/keycard_auth/attack_hand(mob/user)
-	if(user.stat || stat & (NOPOWER|BROKEN))
-		user << "<span class='warning'>This device is not powered!</span>"
-		return
-	if(busy)
-		user << "<span class='warning'>This device is busy!</span>"
-		return
-
-	user.set_machine(src)
-
-	var/dat = "<h1>Keycard Authentication Device</h1>"
-
-	dat += "This device is used to trigger some high security events. It requires the simultaneous swipe of two high-level ID cards."
-	dat += "<br><hr><br>"
-
-	if(screen == 1)
-		dat += "Select an event to trigger:<ul>"
-		dat += "<li><A href='?src=\ref[src];triggerevent=Red alert'>Red alert</A></li>"
-		dat += "<li><A href='?src=\ref[src];triggerevent=Emergency Maintenance Access'>Emergency Maintenance Access</A></li>"
-		dat += "</ul>"
-		user << browse(dat, "window=keycard_auth;size=500x250")
-	if(screen == 2)
-		dat += "Please swipe your card to authorize the following event: <b>[event]</b>"
-		dat += "<p><A href='?src=\ref[src];reset=1'>Back</A>"
-		user << browse(dat, "window=keycard_auth;size=500x250")
 	return
 
+/obj/machinery/keycard_auth/ui_interact(mob/user, ui_key = "main", datum/tgui/ui = null, force_open = 0, \
+					datum/tgui/master_ui = null, datum/ui_state/state = physical_state)

```

Comment:
If state is physical_state, pretty sure AIs cannot interact anyway (just like canisters).